### PR TITLE
vimPlugins.middleclass: build from lua package; vimPlugins.{animation,windows}-nvim: fix deps

### LINKED ISSUE #312050: vimPlugins.windows-nvim: dependency not found
### Describe the bug
One of the dependencies of the plugin, `middleclass`, is not found by the plugin:
```
E5113: Error while calling lua chunk: ...ovimPackages/start/animation.nvim/lua/animation/init.lua:1: module 'middleclass' not found:                                                           
        no field package.preload['middleclass']                                                                                                                                                
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/share/lua/5.1/middleclass.lua'                                                                          
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/share/lua/5.1/middleclass/init.lua'                                                                     
        no file '/nix/store/kf5k5qa53g9rrsch256vqgnr7al080al-luajit-2.1.1693350652-env/lib/lua/5.1/middleclass.so'                                                                             
stack traceback:                                                                                                                                                                               
        [C]: in function 'require'                                                                                                                                                             
        ...ovimPackages/start/animation.nvim/lua/animation/init.lua:1: in main chunk                                                                                                           
        [C]: in function 'require'                                                                                                                                                             
        ...windows.nvim/lua/windows/lib/resize-windows-animated.lua:5: in main chunk                                                                                                           
        [C]: in function 'require'                                                                                                                                                             
        ...myNeovimPackages/start/windows.nvim/lua/windows/init.lua:8: in function 'setup'                                                                                                     
        /home/ozkutuk/.config/nvim/lua/extra.lua:297: in main chunk                                                                                                                            
        [C]: in function 'require'                                                                                                                                                             
        /home/ozkutuk/.config/nvim/init.lua:5: in main chunk
```
It is interesting, because the dependency is actually listed in the overrides:
```nix
  windows-nvim = super.windows-nvim.overrideAttrs {
    dependencies = with self; [ luaPackages.middleclass animation-nvim ];
  };
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.7, NixOS, 24.05 (Uakari), 24.05.20240419.5c24cf2`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.21.2`
 - nixpkgs: `/etc/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by gepbird**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/312050.
Neovim didn't pick up `luaPackages.middleclass` dependency of `windows-nvim` for 2 main reasons (from what I've learned):
- giving a vim plugin a lua package as a dependency doesn't get loaded, it needs to be wrapped by `buildVimPlugin`
- original middleclass had its lua file at the root, but for nvim plugins lua files should be in the lua folder, that's why a fork exists which `vimPlugins.middleclass` uses

Superseeds https://github.com/NixOS/nixpkgs/pull/307026.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by teto** [comment]
this should be fixed by some code in staging-next see https://github.com/NixOS/nixpkgs/issues/318925 . So let's not merge this workaround till the https://github.com/NixOS/nixpkgs/pull/319325 reaches master

**Comment by gepbird** [comment]
I rebased to nixos-unstable where #319325 exists, but unfortunately putting `luaPackages.middleclass` to the dependencies of `windows-nvim` and `animation-nvim` doesn't fix the issue. The reason for this is the original `middleclass` lua package has a different folder structure than what nvim expects, that's why the creator of `windows.nvim` [forked middleclass](https://github.com/anuvyklack/middleclass). I think this PR is fine as it is now.

**Comment by PerchunPak** [comment]
Result of `nixpkgs-review pr 312311` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>vimPlugins.animation-nvim</li>
    <li>vimPlugins.middleclass</li>
    <li>vimPlugins.windows-nvim</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

File: pkgs/applications/editors/vim/plugins/overrides.nix
Comment (PerchunPak) [inline pkgs/applications/editors/vim/plugins/overrides.nix:1013]: How about adding a test?

```suggestion
  middleclass = neovimUtils.buildNeovimPlugin { luaAttr = "middleclass"; nvimRequireCheck = "middleclass"; };
```

Comment (gepbird) [reply pkgs/applications/editors/vim/plugins/overrides.nix]: Thanks for letting me know this exists!

**Review by PerchunPak — APPROVED** [review]
LGTM! Thanks!

[CC](https://github.com/NixOS/nixpkgs/pull/333817#issuecomment-2294798601) @GaetanLepage

**Comment by teto** [comment]
For reference these are now available on luarocks:
https://luarocks.org/modules/neorocks/animation.nvim
https://luarocks.org/modules/neorocks/windows.nvim

It's a bit tough to add luarocks-based neovim plugins at the moment (trying to address that in https://github.com/NixOS/nixpkgs/issues/335931) and since you've spent time on this, I am ok to merge if anyone confirms this makes animation.nvim work . @PerchunPak have you tested it ?

**Comment by PerchunPak** [comment]
Yep, it works
@teto

**Pull request merged**
